File: day03/gear-ratios.py
from collections import defaultdict
from functools import reduce
import re

file = open("input.txt").read().strip()
lines = file.split('\n')

points = [] # what points to check around special chars
dirs = [(-1, -1), (0, -1), (1, -1), (1, 0), (1, 1), (0, 1), (-1, 1), (-1, 0)] # positions to check around special chars

# fill points with points that needed to be checked
for row, line in enumerate(lines):
    for idx, c in enumerate(line):
        if not c.isdigit() and c != '.':
            for dir in dirs:
              points.append((row+dir[0],idx+dir[1]))          
res1 = 0
possible_gear = defaultdict(list)

for row, line in enumerate(lines):
    nums = re.findall(r'(\d+)', line) # find all numbers in the line
    end = 0
    for num in nums:
        idx = line.find(num,end) # get start of num
        end = idx+len(num) # get end of num
        for col in range(idx,idx+len(num)):
            if ((row,col)) in points:
                res1 += int(num) # if number is touching a special char add it
                gear_idx = int(points.index((row,col))/8)
                possible_gear[gear_idx].append(int(num))
                break

gear_ratio = sum(map(lambda x: reduce(lambda y, z: y * z, x[1]) if len(x[1]) > 1 else 0, possible_gear.items())) # get gear ratio

print("part1: ", res1)
print("part2: ", gear_ratio)

File: day03/gear-ratios.js
const fs = require('fs');

// Part 1
const data = fs.readFileSync('input.txt', 'utf-8')
const lines = data.split('\n')

let sum1 = 0
let list = []

for(let i = 0; i < lines.length; i++){
    const norm_chars = '1234567890.'
    let line = lines[i]
    let nums = line.match(/\d+/g)
    let above = '.'.repeat(line.length)
    let below = '.'.repeat(line.length)

    if(i == 0){ // first line
        below = lines[1]
    }else if(i == lines.length -1){ // last line
        above = lines[lines.length - 2]
    }else{ // any line in the middle
        above = lines[i-1]
        below = lines[i+1]
    }

    if(nums){
        let count = 0
        nums.forEach((num) => {
            let valid = false
            let start = line.indexOf(num)
            let end = start + num.length
    
            if(start == 0){ // num at beginning
                end = end + 1
                // check above and below
                for(let j = start; j < end; j++){
                    if(!norm_chars.includes(above[j])) valid = true
                    if(!norm_chars.includes(below[j])) valid = true
                }
                // check right
                if(!norm_chars.includes(line[end - 1])) valid = true
    
            }else if(end == line.length){ // num at end
                start = start - 1
                // check above and below
                for(let j = start; j < end; j++){
                    if(!norm_chars.includes(above[j])) valid = true
                    if(!norm_chars.includes(below[j])) valid = true
                }
                // check left
                if(!norm_chars.includes(line[start])) valid = true
            }else{ // num in middle
                start = start - 1
                end = end + 1
                // check above and below
                for(let j = start; j < end; j++){
                    if(!norm_chars.includes(above[j])) valid = true
                    if(!norm_chars.includes(below[j])) valid = true
                }
                // check right
                if(!norm_chars.includes(line[end - 1])) valid = true
                // check left
                if(!norm_chars.includes(line[start])) valid = true
            }

            if(valid){
                //console.log(above.slice(start, end))
                //console.log(line.slice(start, end))
                //console.log(below.slice(start, end))
            }
    
            if(valid) sum1 += Number(num)
            if(valid) list.push(num)
        })
    }
}


console.log('Part1: ' + sum1)